pkg/extensions/tbtc/recovery/storage.go

```diff
@@ -0,0 +1,106 @@
+package recovery
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+)
+
+const (
+	fileName = "derivation_indexes"
+)
+
+type DerivationIndexStorage struct {
+	path string
+}
+
+func NewDerivationIndexStorage(path string) (*DerivationIndexStorage, error) {
+	err := ensureDirectoryExists(fmt.Sprintf("%s/%s", path, fileName))
+	if err != nil {
+		return nil, err
+	}
+	return &DerivationIndexStorage{
+		path: path,
+	}, nil
+}
+
+func (dis DerivationIndexStorage) getStoragePath(extendedPublicKey string) string {
+	publicKeyDescriptor := extendedPublicKey[:4]
+	suffix := extendedPublicKey[len(extendedPublicKey)-8:]
+	return fmt.Sprintf("%s/%s/%s_%s", dis.path, fileName, publicKeyDescriptor, suffix)
+}
+
+func ensureDirectoryExists(path string) error {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		err = os.Mkdir(path, os.ModePerm)
+		if err != nil {
+			return fmt.Errorf("error occurred while creating a dir: [%v]", err)
+		}
+	}
+
+	return nil
+}
+
+func (dis DerivationIndexStorage) Save(extendedPublicKey string, index int, btcAddress string) error {
+	dirPath := dis.getStoragePath(extendedPublicKey)
+
+	err := ensureDirectoryExists(dirPath)
+	if err != nil {
+		return err
+	}
+	filePath := fmt.Sprintf("%s/%d", dirPath, index)
+
+	return write(filePath, []byte(btcAddress))
+}
+
+func (dis DerivationIndexStorage) Read(extendedPublicKey string) (int, error) {
+	dirPath := dis.getStoragePath(extendedPublicKey)
+
+	err := ensureDirectoryExists(dirPath)
+	if err != nil {
+		return 0, err
+	}
+
+	index := -1
+	files, err := ioutil.ReadDir(dirPath)
+	for _, file := range files {
+		fileIndex, err := strconv.Atoi(file.Name())
+		if err != nil {
+			return 0, err
+		}
+		if fileIndex > index {
+			index = fileIndex
+		}
+	}
+	return index + 1, nil
```

Comment:
I think `+1` in this place is confusing. I'd expect `Read` in the storage to return the last index written to disk and if I want to get the next index that should be used, it should happen in another function, e.g. `GetNextIndex` that will `Read` and `+1`. WDYT?

Reply:
I like the additional clarity